ASoC: SOF: Intel: hda/tgl: Add get_stream_position support

ujfalusi opened the pull request:
Get HDaudio link position for current stream delay calculation from HDA registers, similarly to ACE platforms.

ujfalusi committed:
ASoC: SOF: Intel: hda: Implement get_stream_position (Linear Link Position)

When the Linear Link Position is not available in firmware SRAM window we
use the host accessible position registers to read it.
The address of the PPLCLLPL/U registers depend on the number of streams
(playback+capture).
At probe time the pplc_addr is calculated for each stream and we can use
it to read the LLP without the need of address re-calculation.

Set the get_stream_position callback in sof_hda_common_ops for all
platforms:
The callback is used for IPC4 delay calculations only but the register is
a generic HDA register, not tied to any specific IPC version.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

plbossart commented on sound/soc/sof/intel/tgl.c:
```diff
@@ -106,6 +106,8 @@ int sof_tgl_ops_init(struct snd_sof_dev *sdev)
 
 		sof_tgl_ops.set_power_state = hda_dsp_set_power_state_ipc4;
+
+		sof_tgl_ops.get_stream_position = hda_dsp_get_stream_hda_link_position;
```
I don't know how this can work.

the stream position is always given by DPIB. 
The link position is irrelevant in case of mixing, etc.
We have to rely on the firmware information to know WHEN the stream was added into a mixer and deal with the delay report from the firmware.

ujfalusi replied:
The `get_stream_position` is to get the _link_ position, introduced by
2cd67feff14f ("ASoC: SOF: add get_stream_position ops for pcm delay")
and the MTL implementation is
fe43dab744c0 ("ASoC: SOF: add get_stream_position support on mtl platforms")

LNL inherited this.

ujfalusi replied:
With this patch CAV2.5 and AVS1 behaves in a same way. If the LLP information is not available from firmware then we use the HDA registers.

plbossart replied:
yes we behave in the same way, but how good is it in the first place? You still need the information provided by the firmware to get an offset. Just try with the deep buffer output enabled 10s later than the regular output, this could be fun.

ujfalusi replied:
The start offset is recorded, I have verified it on UPX-i11 HDA to analog headset.

RanderWang reviewed (approved):
LGTM

plbossart commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -1081,3 +1081,18 @@ snd_pcm_uframes_t hda_dsp_stream_get_position(struct hdac_stream *hstream,
+	llp_l = readl(hext_stream->pplc_addr + AZX_REG_PPLCLLPL);
+	llp_u = readl(hext_stream->pplc_addr + AZX_REG_PPLCLLPU);
+
+	return ((u64)llp_u << 32) | llp_l;
```
can someone double-check that if this function is any different from 

````c
u64 mtl_dsp_get_stream_hda_link_position(struct snd_sof_dev *sdev,
					 struct snd_soc_component *component,
					 struct snd_pcm_substream *substream)
{
	struct hdac_stream *hstream = substream->runtime->private_data;
	u32 llp_l, llp_u;

	llp_l = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, MTL_PPLCLLPL(hstream->index));
	llp_u = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR, MTL_PPLCLLPU(hstream->index));
	return ((u64)llp_u << 32) | llp_l;
}
````

HDAudio registers have not really changed between generations, have they?

ujfalusi replied:
@plbossart, the `pplc_addr` is calculated based on the number of playback/capture streams which the `MTL_PPLCLLPL_BASE` is a define which can be working if the number of streams differ. imho the mtl variant needs update to calculate the PPLC base offset, but let me check...

ujfalusi replied:
on TGL:
```
# dmesg | grep peter
[  210.542444] [peter] hda_dsp_stream_init: 0: mtl vs cavs: 0x948 vs 0x118
[  210.542450] [peter] hda_dsp_stream_init: 1: mtl vs cavs: 0x958 vs 0x128
[  210.542456] [peter] hda_dsp_stream_init: 2: mtl vs cavs: 0x968 vs 0x138
[  210.542460] [peter] hda_dsp_stream_init: 3: mtl vs cavs: 0x978 vs 0x148
[  210.542465] [peter] hda_dsp_stream_init: 4: mtl vs cavs: 0x988 vs 0x158
[  210.542469] [peter] hda_dsp_stream_init: 5: mtl vs cavs: 0x998 vs 0x168
[  210.542473] [peter] hda_dsp_stream_init: 6: mtl vs cavs: 0x9a8 vs 0x178
[  210.542479] [peter] hda_dsp_stream_init: 7: mtl vs cavs: 0x9b8 vs 0x188
[  210.542483] [peter] hda_dsp_stream_init: 8: mtl vs cavs: 0x9c8 vs 0x198
[  210.542486] [peter] hda_dsp_stream_init: 9: mtl vs cavs: 0x9d8 vs 0x1a8
[  210.542489] [peter] hda_dsp_stream_init: 10: mtl vs cavs: 0x9e8 vs 0x1b8
[  210.542493] [peter] hda_dsp_stream_init: 11: mtl vs cavs: 0x9f8 vs 0x1c8
[  210.542496] [peter] hda_dsp_stream_init: 12: mtl vs cavs: 0xa08 vs 0x1d8
[  210.542499] [peter] hda_dsp_stream_init: 13: mtl vs cavs: 0xa18 vs 0x1e8
[  210.542503] [peter] hda_dsp_stream_init: 14: mtl vs cavs: 0xa28 vs 0x1f8
[  210.542506] [peter] hda_dsp_stream_init: 15: mtl vs cavs: 0xa38 vs 0x208
```
```diff
diff --git a/sound/soc/sof/intel/hda-stream.c b/sound/soc/sof/intel/hda-stream.c
index 125dc586439b..e9075dc846c7 100644
--- a/sound/soc/sof/intel/hda-stream.c
+++ b/sound/soc/sof/intel/hda-stream.c
@@ -22,6 +22,7 @@
 #include "../ops.h"
 #include "../sof-audio.h"
 #include "hda.h"
+#include "mtl.h"
 
 int sof_hda_position_quirk = SOF_HDA_POSITION_QUIRK_USE_DPIB_REGISTERS;
 module_param_named(position_quirk, sof_hda_position_quirk, int, 0444);
@@ -936,6 +937,11 @@ int hda_dsp_stream_init(struct snd_sof_dev *sdev)
 		hstream->opened = false;
 		hstream->running = false;
 
+		pr_warn("[peter] %s: %d: mtl vs cavs: %#x vs %#x\n", __func__, i,
+			MTL_PPLCLLPL(hstream->index),
+			SOF_HDA_PPLC_BASE + SOF_HDA_PPLC_MULTI * num_total +
+			SOF_HDA_PPLC_INTERVAL * i + AZX_REG_PPLCLLPL);
+
 		if (i < num_capture) {
 			hstream->stream_tag = i + 1;
 			hstream->direction = SNDRV_PCM_STREAM_CAPTURE;
```

plbossart replied:
In the HDaudio ext stream stuff, there is nothing that's SOC dependent

````
	hext_stream->pplc_addr = bus->ppcap + AZX_PPLC_BASE +
				AZX_PPLC_MULTI * bus->num_streams +
				AZX_PPLC_INTERVAL * idx;
````

I don't buy the argument that MTL registers are different.

ujfalusi replied:
fwiw, on MTL the same diff:
```
# dmesg -w | grep peter
[    5.249721] [peter] hda_dsp_stream_init: 0: mtl vs cavs: 0x948 vs 0x148
[    5.249726] [peter] hda_dsp_stream_init: 1: mtl vs cavs: 0x958 vs 0x158
[    5.249729] [peter] hda_dsp_stream_init: 2: mtl vs cavs: 0x968 vs 0x168
[    5.249731] [peter] hda_dsp_stream_init: 3: mtl vs cavs: 0x978 vs 0x178
[    5.249734] [peter] hda_dsp_stream_init: 4: mtl vs cavs: 0x988 vs 0x188
[    5.249736] [peter] hda_dsp_stream_init: 5: mtl vs cavs: 0x998 vs 0x198
[    5.249738] [peter] hda_dsp_stream_init: 6: mtl vs cavs: 0x9a8 vs 0x1a8
[    5.249741] [peter] hda_dsp_stream_init: 7: mtl vs cavs: 0x9b8 vs 0x1b8
[    5.249743] [peter] hda_dsp_stream_init: 8: mtl vs cavs: 0x9c8 vs 0x1c8
[    5.249745] [peter] hda_dsp_stream_init: 9: mtl vs cavs: 0x9d8 vs 0x1d8
[    5.249747] [peter] hda_dsp_stream_init: 10: mtl vs cavs: 0x9e8 vs 0x1e8
[    5.249750] [peter] hda_dsp_stream_init: 11: mtl vs cavs: 0x9f8 vs 0x1f8
[    5.249752] [peter] hda_dsp_stream_init: 12: mtl vs cavs: 0xa08 vs 0x208
[    5.249754] [peter] hda_dsp_stream_init: 13: mtl vs cavs: 0xa18 vs 0x218
[    5.249756] [peter] hda_dsp_stream_init: 14: mtl vs cavs: 0xa28 vs 0x228
[    5.249758] [peter] hda_dsp_stream_init: 15: mtl vs cavs: 0xa38 vs 0x238
[    5.249761] [peter] hda_dsp_stream_init: 16: mtl vs cavs: 0xa48 vs 0x248
[    5.249763] [peter] hda_dsp_stream_init: 17: mtl vs cavs: 0xa58 vs 0x258
[    5.249765] [peter] hda_dsp_stream_init: 18: mtl vs cavs: 0xa68 vs 0x268
```
There is 0x800 offset between the two address, I need to find one MTL HDA device to see what the actual readings will return.

RanderWang replied:
I check the spec again and confirm that 0x948 is for LLPL, 0x94c is for LLPU.

RanderWang replied:
I checked the MTL and get this log, ppl offset is at 0x800.  Above log should add this offset.
Using method in this PR can unify the MTL & TGL , so @ujfalusi  can do more work

````
[71383.866630] :hda_dsp_ctrl_get_caps:  0000:00:1f.3: found DSP capability at 0x800
````

ujfalusi replied:
Right, I'm blind.
The registers that we read is the same in both cases:
hda-ctrl.c:
```c
dev_dbg(sdev->dev, "found DSP capability at 0x%x\n", offset);
bus->ppcap = bus->remap_addr + offset;
sdev->bar[HDA_DSP_PP_BAR] = bus->ppcap;
```
-> found DSP capability at **0x800**

hda-stream.c:
```c
hext_stream->pplc_addr = sdev->bar[HDA_DSP_PP_BAR] +
			 SOF_HDA_PPLC_BASE + SOF_HDA_PPLC_MULTI * num_total +
			 SOF_HDA_PPLC_INTERVAL * i;
```
Offset for stream 0:
TGL: 0x118 (detected 9 playback and 7 capture streams)
MTL: 0x148 (detected 9 playback and 10 capture streams)

```c
#define MTL_PPLCLLPL_BASE		0x948
```
So, it is hardwired to 0x800+0x148

I'll update the PR to use the `hda_dsp_get_stream_hda_link_position()` on all platforms, drop the mtl implementation (which likely not work on MTL variants or on LNL):
```c
llp_l = readl(hext_stream->pplc_addr + AZX_REG_PPLCLLPL);
llp_u = readl(hext_stream->pplc_addr + AZX_REG_PPLCLLPU);
```

RanderWang replied:
@ujfalusi  use snd_hdac_chip_readl(bus, PPLCLLPL); ? we use it in hda-codec.c

RanderWang replied:
sorry, it doesn't work. Ignore it

ujfalusi committed:
ASoC: SOF: Intel: mtl/lnl: Use the generic get_stream_position callback

Drop the MTL mtl_dsp_get_stream_hda_link_position() function and related
defines since it can only work on platforms which have 19 streams because
of the use of 0x948 as base offset for the LLP registers.

The generic hda_dsp_get_stream_hda_link_position() takes the number of
streams into consideration when reading the LLP registers for the stream
and can handle different HDA configurations.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v1:
- Add small comment section in hda_dsp_get_stream_hda_link_position()
- set  the get_stream_position callback unconditionally to hda_dsp_get_stream_hda_link_position()
- Drop the MTL LLP reading implementation and use the generic one

plbossart commented on sound/soc/sof/intel/hda-common-ops.c:
```diff
@@ -57,6 +57,8 @@ struct snd_sof_dsp_ops sof_hda_common_ops = {
 	.pcm_ack	= hda_dsp_pcm_ack,
 
+	.get_stream_position = hda_dsp_get_stream_hda_link_position,
+
```
typo in commit message: s/regsiter/register/

plbossart reviewed (approved):
Nice clean-up, thank you @ujfalusi

ujfalusi commented:
Changes since v2:
- fixed typos in the commit message of the first patch

RanderWang reviewed (approved):
LGTM

ujfalusi commented:
Moved to draft and back to ready:
The delay reporting is not correct but TGL is not officially supported, so this PR will not introduce regression, but fix the ACE variants.
